In [1]:
import numpy as np
from matplotlib import pyplot as plt

file_data = open('data/Assignment1_Q2_Data.csv') 


data = []
count = 0
for i in file_data:
    if count != 0:
        x = i.split(",")
        data.append(x)
    count += 1


Land_area = []
for i in data:
    Land_area.append(i[3])

Living_area = []
for i in data:
    Living_area.append(i[4])

price = []
for i in data:
    price.append(i[12])

for i in range(len(price)-1):
    price[i] = price[i][ :-1 ]

bathrooms = []
for i in data:
    bathrooms.append(i[2])
bathrooms = np.array(bathrooms, dtype='float64')
bathrooms = bathrooms.reshape(1, -1)

garages = []
for i in data:
    garages.append(i[5])
garages = np.array(garages, dtype='float64')
garages = garages.reshape(1, -1)

rooms = []
for i in data:
    rooms.append(i[6])
rooms = np.array(rooms, dtype='float64')
rooms = rooms.reshape(1, -1)

bedroom = []
for i in data:
    bedroom.append(i[7])
bedroom = np.array(bedroom, dtype='float64')
bedroom = bedroom.reshape(1, -1)

age_of_home = []
for i in data:
    age_of_home.append(i[8])
age_of_home = np.array(age_of_home, dtype='float64')
age_of_home = age_of_home.reshape(1, -1)

cons = []
for i in data:
    cons.append(i[9])
cons = np.array(cons, dtype='float64')
cons = cons.reshape(1, -1)

arch = []
for i in data:
    arch.append(i[10])
arch = np.array(arch, dtype='float64')
arch = arch.reshape(1, -1)

fire_place = []
for i in data:
    fire_place.append(i[11])
fire_place = np.array(fire_place, dtype='float64')
fire_place = fire_place.reshape(1, -1)


Land_area = np.array(Land_area, dtype='float64')
Living_area = np.array(Living_area, dtype='float64')
price = np.array(price, dtype='float64')

Land_area = Land_area.reshape(1, -1)
Living_area = Living_area.reshape(1, -1)

x_train = np.concatenate((Land_area, Living_area, bathrooms, garages, rooms, bedroom, age_of_home, cons, arch, fire_place), axis=0)
x_train = x_train.swapaxes(0, 1)
y_train = price

X = x_train
Y = y_train

weight = np.zeros((1, x_train.shape[1]))
bias = 0.0
LR = 0.02
epochs = 360000

In [2]:
# training loop
for i in range(epochs): 

    # make prediction
    Y_pred = np.matmul(X, weight.T) + bias

    # calcuate weight and bias update
    weight_update = (-2/(1000)) * np.dot( (Y - Y_pred).T, X) 
    bias_update = (-2/(1000)) * np.sum(Y - Y_pred) 

    # add updates to weights and bias
    weight = weight - LR * weight_update  # Update m
    bias = bias - LR * bias_update  # Update c

    # calacute loss 
    MSE1 = np.square(np.subtract(y_train, Y_pred)).mean()
    
    # testing data predictions and loss

In [3]:
print("train loss:", MSE1, "epoch:", i)

train loss: 1.689694796165722 epoch: 359999


1. What is the average least squares error for the given data using your simple linear regression
model? 

Answer: 1.689

2. Which factor has the most effect on the final value? How do you know this? Can you use only
this feature to predict the price?

Answer: Living room, because I took every feature out and when this feature was out I got the highest MSE. And no you can not use only living room for predicting the price. The MSE was 21.72

3. Which factor has the least effect on the final value? How do you know this? What effect does
removing this feature have on the performance?

Answer: Number of Bedroom, when I took this feature out the model still had the lowest MSE out of the other features. It has almost no effect with a MSE of 1.69 when its took out vs 1.689 when it's in.
